Reduce the pre-image flood

linked0 opened the pull request:
We previously took a rather aggressive approach to pre-image
propagation, which leads to a lot of unnecessary traffic. So
We check that the pre-images to be sent are already posted in
order to prevent the too heavy traffic.

Fixes #2356

hewison-chris reviewed (approved):
LGTM

linked0 commented:
Ready for review with all green.

Geod24 commented on source/agora/network/Client.d:
```diff
@@ -223,7 +226,19 @@ public class NetworkClient
 
         case Preimage:
-            this.attemptRequest!(API.postPreimage, Throw.No)(event.preimage);
+            // Clean up the `last_preimages` priodically
+            // Note: It clears `last_preimages` everyday based on the CoinNet.
+            if (event.preimage.height % Height(144) == 0)
+                this.last_preimages.clear();
```
`periodically`

Geod24 commented on source/agora/network/Client.d:
```diff
@@ -223,7 +226,19 @@ public class NetworkClient
+            // We do not post the preimage if it is already posted in order to
+            // prevent calls of `postPreimage` from flooding
+            if (event.preimage.utxo !in this.last_preimages ||  
+                event.preimage.height > this.last_preimages[event.preimage.utxo].height)
```
We only need to store the `Height` actually, not the full `PreImageInfo`, right ?

Geod24 commented on source/agora/test/PreImageCatchup.d:
```diff
@@ -0,0 +1,64 @@
+import agora.consensus.data.PreImageInfo;
+import agora.test.Base;
+
+/// A validator should retrive the pre-images even when the `postPreimage`
```
`retrieve`

Geod24 commented on source/agora/test/PreImageCatchup.d:
```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+
+    Check that a validator can retrieve missing pre-images when it could not
+    receive pre-images through `postPreimage`. We altered the algorithm of
+    distributing pre-images so that a validator does not continuously send
+    pre-images any more without checking that the previous pre-image is
+    posted normally. So the network should survive the situation where
+    some pre-images are missing.
```
Small DDOC note: The title/summary should be short, and the longer explanation should follow.
E.g.
```D
/*******************************************************************************

    Some high level overview about this, in a single sentence, if possible.

    A longer explanation that can span as many sentences as one want.
    And obviously multiple line. Give as much details as you wish, really.

    Specific_notes:
      For specific topic, it's possible to have sub-heading using the same
      approach as `Params` and `Returns`. In fact, `Params`, `Returns`,
      and the likes are simply recognized sub heading and don't need to come
      in a specific order.

*******************************************************************************/
```

linked0 replied:
Simplified the summary of the unit test because the goal of the test has been made clearer.

Geod24 commented on source/agora/test/PreImageCatchup.d:
```diff
@@ -0,0 +1,64 @@
+/// does not work or the validator misses some pre-images.
+unittest
+{
+    static class CustomValidator : TestValidatorNode
+    {
+        mixin ForwardCtor!();
+
+        /// It makes the situation where the node doesn't receive pre-images
+        /// and miss some preimages from some point.
+        public override bool postPreimage (in PreImageInfo preimage) @safe
+        {
+            if (this.ledger.getBlockHeight() <= 1)
+                return super.postPreimage(preimage);
+            return true;
+        }
+    }
```
What is this testing ? The golden rule of testing is that your test should fail if the fix is not present.
From the look of it, this test would pass even without the fix.

linked0 replied:
Modified the test based on the comment.

Geod24 reviewed (changes requested):
Test is lacking

linked0 commented:
> Test is lacking

Thanks for the review. I will make a unit test more proper. Anyway, do you think this approach for keeping pre-images from flooding is right?

Geod24 commented:
Yes

linked0 committed:
Add member related to pre-image to `TestConf`

We add `preimage_reveal_interval` and `max_preimage_reveal` to TestConf.
It's because we need to refer to the variables in order to execute unit
tests more accurately.

linked0 committed:
Reduce the pre-image flood

We previously took a rather aggressive approach to pre-image
propagation, which leads to a lot of unnecessary traffic. So
We check that the pre-images to be sent is already posted in
order to prevent the too heavy traffic.

hewison-chris commented on source/agora/test/PreImageNoFlood.d:
```diff
@@ -0,0 +1,84 @@
+
+    // The count of calling `postPreimage` does not change after a while
+    Thread.sleep(config.preimage_reveal_interval);
+    auto prev_post_count = atomicLoad(CustomAPIManager.post_preimage_count);
```
Can we not check the expected count is the number of other validators here?

linked0 replied:
I understand what you mean but I found that the expected count changes every time in the middle of running the unit test with checking the count.

hewison-chris reviewed (changes requested):
I think it would be better to check the exact number of posts.
Also, the test you previously had for pre-image catchup is maybe worth adding back as a commit if it is not already covered by another test.

linked0 commented:
> I think it would be better to check the exact number of posts.

I don't think it's meaningful because I found that the number of posts could be changed with just a little change in the environment, like some unintended delay based on machines on which it is tested.

> Also, the test you previously had for pre-image catchup is maybe worth adding back as a commit if it is not already covered by another test.

I agree with you. Will add it again.

linked0 commented:
Applied suggestions to the code.

hewison-chris reviewed (approved):
LGTM

Geod24 commented on source/agora/network/Client.d:
```diff
@@ -156,6 +156,9 @@ public class NetworkClient
     /// Gossip delay
     private enum GossipDelay = 10.msecs;
 
+    /// Heights of last gossiped pre-images
```
Mention that the key is the stake

linked0 replied:
Will change to `/// Heights of last gossiped pre-images for staked UTXOs`. If it's not proper, please let me know.

Geod24 commented on source/agora/network/Client.d:
```diff
@@ -227,7 +230,20 @@ public class NetworkClient
+
+            // We do not post the preimage if it is already posted in order to
+            // prevent calls of `postPreimage` from flooding
+            if (event.preimage.utxo !in this.last_preimages ||
+                event.preimage.height > this.last_preimages[event.preimage.utxo])
+            {
+                auto height = this.attemptRequest!(API.postPreimage, Throw.No)(event.preimage);
+                if (height >= event.preimage.height)
+                    this.last_preimages[event.preimage.utxo] = event.preimage.height;
+            }
```
```suggestion
            if (event.preimage.height > this.last_preimages.get(event.preimage.utxo, Height(0)))
            {
                auto height = this.attemptRequest!(API.postPreimage, Throw.No)(event.preimage);
                if (height >= event.preimage.height)
                    this.last_preimages[event.preimage.utxo] = event.preimage.height;
            }
```

linked0 replied:
The `get` function is not `nothrow`. Isn't it better to leave as it is?

Geod24 reviewed (approved):
Just a couple style changes, then feel free to self-merge
